okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/QosAwareCallRetryerTest.java

```diff
@@ -0,0 +1,340 @@
+
+    class BlockingCallVerifier implements RetryVerifier {
+
+        @Override
+        public void verifyBackoffAndRetry(IOException exception, int numRetries) throws Exception {
+            Response response = call.execute();
+
+            verify(sleeper, times(numRetries)).sleepForBackoff(BACKOFF_10SECS.get());
+            verify(delegateCall, times(1)).execute();
+
+            if (numRetries > 0) {
+                verify(delegateCall, times(1)).clone();
+                verify(clonedDelegateCall, times(numRetries)).execute();
+                verify(clonedDelegateCall, times(numRetries - 1)).clone();
+            }
+
+            verifyNoMoreInteractions(delegateCall, clonedDelegateCall);
+
+            assertThat(response).isEqualTo(numRetries > 0 ? CLONED_RESPONSE : RESPONSE);
+        }
+
+        @Override
+        public AbstractThrowableAssert<?, ? extends Throwable> verifyNoRetryAndThrow(QosIoException exception)
+                throws Exception {
+            AbstractThrowableAssert<?, ? extends Throwable> thrown =
+                    assertThatThrownBy(() -> retrier.executeWithRetry(call));
+
+            verify(delegateCall).execute();
+            verifyZeroInteractions(sleeper);
+
+            verifyNoMoreInteractions(delegateCall, clonedDelegateCall);
+
+            return thrown;
+        }
+    }
+
+    class AsyncCallVerifier implements RetryVerifier {
+
+        @Override
+        public void verifyBackoffAndRetry(IOException exception, int numRetries) {
+            FutureCallback response = new FutureCallback();
+            call.enqueue(response);
+
+            if (numRetries > 0) {
+                scheduler.tick(numRetries * 10 - 1, TimeUnit.SECONDS);
+                assertThat(response.isDone()).isFalse();
+                assertThat(response.isCancelled()).isFalse();
+            }
+
+            scheduler.tick(1, TimeUnit.NANOSECONDS);
+            scheduler.tick(10 * numRetries, TimeUnit.SECONDS);
+            assertThat(response.isDone()).isTrue();
+            assertThat(response.getNow(null)).isEqualTo(numRetries > 0 ? CLONED_RESPONSE : RESPONSE);
+
+            verify(delegateCall, times(1)).enqueue(any());
+            verify(clonedDelegateCall, times(numRetries)).enqueue(any());
+
+            if (numRetries > 0) {
+                verify(delegateCall, times(1)).clone();
+                verify(clonedDelegateCall, times(numRetries - 1)).clone();
+            }
+
+            verifyNoMoreInteractions(delegateCall, clonedDelegateCall);
+        }
+
+        @Override
+        public AbstractThrowableAssert<?, ? extends Throwable> verifyNoRetryAndThrow(QosIoException exception) {
+            FutureCallback response = new FutureCallback();
+            call.enqueue(response);
+            scheduler.tick(100, TimeUnit.SECONDS);
+            assertThat(response.isDone()).isTrue();
+
+            verify(delegateCall, times(1)).enqueue(any());
+
+            verifyNoMoreInteractions(clonedDelegateCall);
```

Comment:
Note to self: figure out why this fails if I include 'delegateCall' in this verify